## Exploratory Data Analysis
>This notebook shows some basic Exploratory Data Analysis about the dataset

### Objectives
0. Clean data
1. Which book is the most popular?
2. Which author is the most popular?
3. Which number wrote the biggest number of books?
4. Is number of pages correlated with ratings or number of reviews?
5. Which years had the biggest number of books written?
6. Is there tendency to reduce number of pages in nowaday books?

### Import libraries

In [1]:
import pandas as pd 
import polars as pl 
import numpy as np 
import matplotlib.pyplot as plt 
import seaborn as sns 
%matplotlib inline
sns.set_style('whitegrid')

### Load the data

In [2]:
books_df = pl.DataFrame(schema=pl.read_csv('dataset/book1000k-1100k.csv').schema)
books_df.head()

shape: (0, 20)
┌─────┬──────┬─────────┬──────┬───┬──────────┬─────────────┬─────────────┬───────────────────────┐
│ Id  ┆ Name ┆ Authors ┆ ISBN ┆ … ┆ Language ┆ pagesNumber ┆ Description ┆ Count of text reviews │
│ --- ┆ ---  ┆ ---     ┆ ---  ┆   ┆ ---      ┆ ---         ┆ ---         ┆ ---                   │
│ i64 ┆ str  ┆ str     ┆ str  ┆   ┆ str      ┆ i64         ┆ str         ┆ i64                   │
╞═════╪══════╪═════════╪══════╪═══╪══════════╪═════════════╪═════════════╪═══════════════════════╡
└─────┴──────┴─────────┴──────┴───┴──────────┴─────────────┴─────────────┴───────────────────────┘

In [17]:
books_df.replace('PublishDay', books_df['PublishDay'].shrink_dtype())
books_df.replace('PublishMonth', books_df['PublishMonth'].shrink_dtype())
books_df.replace('PublishYear', books_df['PublishYear'].shrink_dtype())
books_df.schema

{'Id': Int64,
 'Name': Utf8,
 'Authors': Utf8,
 'ISBN': Utf8,
 'Rating': Float64,
 'PublishYear': Int32,
 'PublishMonth': Int8,
 'PublishDay': Int8,
 'Publisher': Utf8,
 'RatingDist5': Utf8,
 'RatingDist4': Utf8,
 'RatingDist3': Utf8,
 'RatingDist2': Utf8,
 'RatingDist1': Utf8,
 'RatingDistTotal': Utf8,
 'CountsOfReview': Int64,
 'Language': Utf8,
 'pagesNumber': Int64,
 'Description': Utf8,
 'Count of text reviews': Int64,
 'PagesNumber': Int64}

In [18]:
books_df[['PublishDay', 'PublishMonth', 'PublishYear']].describe()

shape: (7, 4)
┌────────────┬────────────┬──────────────┬─────────────┐
│ describe   ┆ PublishDay ┆ PublishMonth ┆ PublishYear │
│ ---        ┆ ---        ┆ ---          ┆ ---         │
│ str        ┆ f64        ┆ f64          ┆ f64         │
╞════════════╪════════════╪══════════════╪═════════════╡
│ count      ┆ 1.850198e6 ┆ 1.850198e6   ┆ 1.850198e6  │
│ null_count ┆ 0.0        ┆ 0.0          ┆ 0.0         │
│ mean       ┆ 7.987328   ┆ 7.700395     ┆ 1997.841519 │
│ std        ┆ 8.431971   ┆ 7.756715     ┆ 87.897239   │
│ min        ┆ 1.0        ┆ 1.0          ┆ 1.0         │
│ max        ┆ 31.0       ┆ 31.0         ┆ 65535.0     │
│ median     ┆ 5.0        ┆ 6.0          ┆ 2000.0      │
└────────────┴────────────┴──────────────┴─────────────┘

- Min and max year looks strange

- Max month is 31, also, mean of day seems strange

In [19]:
books_df['PublishYear'].unique()

shape: (199,)
Series: 'PublishYear' [i32]
[
	1
	8
	162
	199
	200
	202
	208
	299
	1192
	1376
	1384
	1385
	…
	2103
	2202
	2994
	3002
	3006
	4989
	19769
	20015
	20016
	20040
	20067
	20099
	65535
]

- Year 1, 8, 162, 199, 200, 202, 208, 299, 2030, 2035, ... are wrong

- Year 2022 and 2021 need further inspecting

In [20]:
books_df['PublishYear'].value_counts()

shape: (199, 2)
┌─────────────┬────────┐
│ PublishYear ┆ counts │
│ ---         ┆ ---    │
│ i32         ┆ u32    │
╞═════════════╪════════╡
│ 1376        ┆ 1      │
│ 1952        ┆ 145    │
│ 1896        ┆ 2      │
│ 1920        ┆ 145    │
│ …           ┆ …      │
│ 1929        ┆ 35     │
│ 1977        ┆ 10494  │
│ 1865        ┆ 3      │
│ 1937        ┆ 50     │
└─────────────┴────────┘

Let's just take the books from 1700 to 2021 because others seem suspicious

In [21]:
books_df = books_df.sort('PublishYear')

In [22]:
low = books_df.filter(books_df['PublishYear'] < 1700).shape[0]
high = books_df.filter(books_df['PublishYear'] > 2021).shape[0]

In [23]:
books_df = books_df[low:len(books_df)-high]

In [24]:
books_df['PublishYear'].value_counts()

shape: (167, 2)
┌─────────────┬────────┐
│ PublishYear ┆ counts │
│ ---         ┆ ---    │
│ i32         ┆ u32    │
╞═════════════╪════════╡
│ 1730        ┆ 1      │
│ 1753        ┆ 2      │
│ 1824        ┆ 1      │
│ 1825        ┆ 1      │
│ …           ┆ …      │
│ 2018        ┆ 1077   │
│ 2019        ┆ 631    │
│ 2020        ┆ 305    │
│ 2021        ┆ 47     │
└─────────────┴────────┘

#### Publisher

In [25]:
books_df.filter(books_df['Publisher'].is_null()).head()

shape: (5, 21)
┌─────────┬───────────┬───────────┬───────────┬───┬───────────┬───────────┬────────────┬───────────┐
│ Id      ┆ Name      ┆ Authors   ┆ ISBN      ┆ … ┆ pagesNumb ┆ Descripti ┆ Count of   ┆ PagesNumb │
│ ---     ┆ ---       ┆ ---       ┆ ---       ┆   ┆ er        ┆ on        ┆ text       ┆ er        │
│ i64     ┆ str       ┆ str       ┆ str       ┆   ┆ ---       ┆ ---       ┆ reviews    ┆ ---       │
│         ┆           ┆           ┆           ┆   ┆ i64       ┆ str       ┆ ---        ┆ i64       │
│         ┆           ┆           ┆           ┆   ┆           ┆           ┆ i64        ┆           │
╞═════════╪═══════════╪═══════════╪═══════════╪═══╪═══════════╪═══════════╪════════════╪═══════════╡
│ 3088068 ┆ Godey's   ┆ Various   ┆ 142648437 ┆ … ┆ null      ┆ This is a ┆ null       ┆ 220       │
│         ┆ Lady's    ┆           ┆ 2         ┆   ┆           ┆ pre-1923  ┆            ┆           │
│         ┆ Book:     ┆           ┆           ┆   ┆           ┆ historica ┆            ┆           │
│         ┆ Vol. 42   ┆           ┆           ┆   ┆           ┆ l re…     ┆            ┆           │
│         ┆ Jan…      ┆           ┆           ┆   ┆           ┆           ┆            ┆           │
│ 1424633 ┆ An Appeal ┆ Alfred H. ┆ 142975357 ┆ … ┆ 24        ┆ null      ┆ 0          ┆ null      │
│         ┆ in vindic ┆ Love      ┆ 9         ┆   ┆           ┆           ┆            ┆           │
│         ┆ ation of  ┆           ┆           ┆   ┆           ┆           ┆            ┆           │
│         ┆ peac…     ┆           ┆           ┆   ┆           ┆           ┆            ┆           │
│ 2619182 ┆ French    ┆ Jean-Paul ┆ 066920083 ┆ … ┆ null      ┆ null      ┆ null       ┆ 480       │
│         ┆ for       ┆ Valette   ┆ 2         ┆   ┆           ┆           ┆            ┆           │
│         ┆ Mastery 2 ┆           ┆           ┆   ┆           ┆           ┆            ┆           │
│         ┆ - Tous    ┆           ┆           ┆   ┆           ┆           ┆            ┆           │
│         ┆ ense…     ┆           ┆           ┆   ┆           ┆           ┆            ┆           │
│ 2802465 ┆ Foundatio ┆ McDougal  ┆ 066940363 ┆ … ┆ null      ┆ null      ┆ null       ┆ 544       │
│         ┆ ns of Seq ┆ Littell   ┆ 6         ┆   ┆           ┆           ┆            ┆           │
│         ┆ uential   ┆           ┆           ┆   ┆           ┆           ┆            ┆           │
│         ┆ Mathem…   ┆           ┆           ┆   ┆           ┆           ┆            ┆           │
│ 194155  ┆ Dime Uno  ┆ Fabián A. ┆ 066943347 ┆ … ┆ 0         ┆ null      ┆ null       ┆ null      │
│         ┆ Cuaderno  ┆ Samaniego ┆ 0         ┆   ┆           ┆           ┆            ┆           │
│         ┆ De Activi ┆           ┆           ┆   ┆           ┆           ┆            ┆           │
│         ┆ dades     ┆           ┆           ┆   ┆           ┆           ┆            ┆           │
└─────────┴───────────┴───────────┴───────────┴───┴───────────┴───────────┴────────────┴───────────┘

In [26]:
books_df['Publisher'].n_unique()

79419

A lot of books. Also books with good ratings, I should not remove them

In [27]:
# Which publisher issued the biggest variety of books
books_df['Publisher'].value_counts()

shape: (79419, 2)
┌────────────────────────┬────────┐
│ Publisher              ┆ counts │
│ ---                    ┆ ---    │
│ str                    ┆ u32    │
╞════════════════════════╪════════╡
│ Vintage Press          ┆ 1      │
│ Carraig Books          ┆ 1      │
│ Grove's Dictionaries   ┆ 4      │
│ Éditions La Découverte ┆ 3      │
│ …                      ┆ …      │
│ W W Norton & Co Inc    ┆ 9      │
│ Arrow Books Ltd.       ┆ 11     │
│ Watermill Press        ┆ 11     │
│ Becker Press           ┆ 3      │
└────────────────────────┴────────┘

- Missing value: Not Avail, Unknown, Not Specified, Not Applicable, ...

#### RatingDist (1, 2, 3, 4, 5, total)

In [28]:
books_df.head(3)

shape: (3, 21)
┌─────────┬───────────┬───────────┬───────────┬───┬───────────┬───────────┬────────────┬───────────┐
│ Id      ┆ Name      ┆ Authors   ┆ ISBN      ┆ … ┆ pagesNumb ┆ Descripti ┆ Count of   ┆ PagesNumb │
│ ---     ┆ ---       ┆ ---       ┆ ---       ┆   ┆ er        ┆ on        ┆ text       ┆ er        │
│ i64     ┆ str       ┆ str       ┆ str       ┆   ┆ ---       ┆ ---       ┆ reviews    ┆ ---       │
│         ┆           ┆           ┆           ┆   ┆ i64       ┆ str       ┆ ---        ┆ i64       │
│         ┆           ┆           ┆           ┆   ┆           ┆           ┆ i64        ┆           │
╞═════════╪═══════════╪═══════════╪═══════════╪═══╪═══════════╪═══════════╪════════════╪═══════════╡
│ 3098402 ┆ Three     ┆ Samuel    ┆ 141916310 ┆ … ┆ null      ┆ null      ┆ null       ┆ 80        │
│         ┆ Distinct  ┆ Pritchard ┆ 8         ┆   ┆           ┆           ┆            ┆           │
│         ┆ Knocks On ┆           ┆           ┆   ┆           ┆           ┆            ┆           │
│         ┆ the Doo…  ┆           ┆           ┆   ┆           ┆           ┆            ┆           │
│ 2448024 ┆ Discoveri ┆ Michael   ┆ 155953200 ┆ … ┆ null      ┆ null      ┆ null       ┆ 834       │
│         ┆ ng        ┆ Serra     ┆ 9         ┆   ┆           ┆           ┆            ┆           │
│         ┆ Geometry: ┆           ┆           ┆   ┆           ┆           ┆            ┆           │
│         ┆ An        ┆           ┆           ┆   ┆           ┆           ┆            ┆           │
│         ┆ Investi…  ┆           ┆           ┆   ┆           ┆           ┆            ┆           │
│ 4265642 ┆ Self      ┆ Lila      ┆ 084036013 ┆ … ┆ null      ┆ null      ┆ null       ┆ 170       │
│         ┆ Esteem In ┆ Swell     ┆ 4         ┆   ┆           ┆           ┆            ┆           │
│         ┆ The Class ┆           ┆           ┆   ┆           ┆           ┆            ┆           │
│         ┆ room: Te… ┆           ┆           ┆   ┆           ┆           ┆            ┆           │
└─────────┴───────────┴───────────┴───────────┴───┴───────────┴───────────┴────────────┴───────────┘

Get rid of redundant parts like '5:', '4:', 'total:', ...

In [29]:
books_df = books_df.with_columns(books_df['RatingDist1'].apply(lambda rating: rating.split(':')[1]).cast(pl.Int32))
books_df = books_df.with_columns(books_df['RatingDist2'].apply(lambda rating: rating.split(':')[1]).cast(pl.Int32))
books_df = books_df.with_columns(books_df['RatingDist3'].apply(lambda rating: rating.split(':')[1]).cast(pl.Int32))
books_df = books_df.with_columns(books_df['RatingDist4'].apply(lambda rating: rating.split(':')[1]).cast(pl.Int32))
books_df = books_df.with_columns(books_df['RatingDist5'].apply(lambda rating: rating.split(':')[1]).cast(pl.Int32))
books_df = books_df.with_columns(books_df['RatingDistTotal'].apply(lambda rating: rating.split(':')[1]).cast(pl.Int32))
books_df[['RatingDistTotal', 'RatingDist1', 'RatingDist2', 'RatingDist3', 'RatingDist4', 'RatingDist5']].describe()

shape: (7, 7)
┌────────────┬────────────┬─────────────┬─────────────┬──────────────┬──────────────┬──────────────┐
│ describe   ┆ RatingDist ┆ RatingDist1 ┆ RatingDist2 ┆ RatingDist3  ┆ RatingDist4  ┆ RatingDist5  │
│ ---        ┆ Total      ┆ ---         ┆ ---         ┆ ---          ┆ ---          ┆ ---          │
│ str        ┆ ---        ┆ f64         ┆ f64         ┆ f64          ┆ f64          ┆ f64          │
│            ┆ f64        ┆             ┆             ┆              ┆              ┆              │
╞════════════╪════════════╪═════════════╪═════════════╪══════════════╪══════════════╪══════════════╡
│ count      ┆ 1.850149e6 ┆ 1.850149e6  ┆ 1.850149e6  ┆ 1.850149e6   ┆ 1.850149e6   ┆ 1.850149e6   │
│ null_count ┆ 0.0        ┆ 0.0         ┆ 0.0         ┆ 0.0          ┆ 0.0          ┆ 0.0          │
│ mean       ┆ 4079.61434 ┆ 94.402791   ┆ 207.246389  ┆ 754.638235   ┆ 1305.846063  ┆ 1717.480868  │
│            ┆ 6          ┆             ┆             ┆              ┆              ┆              │
│ std        ┆ 71662.0828 ┆ 2227.034126 ┆ 3567.509688 ┆ 11238.758305 ┆ 20767.199856 ┆ 37288.463405 │
│            ┆ 66         ┆             ┆             ┆              ┆              ┆              │
│ min        ┆ -2.0       ┆ 0.0         ┆ 0.0         ┆ 0.0          ┆ 0.0          ┆ -2.0         │
│ max        ┆ 7.094687e6 ┆ 550388.0    ┆ 544093.0    ┆ 1.013165e6   ┆ 1.912159e6   ┆ 4.608992e6   │
│ median     ┆ 5.0        ┆ 0.0         ┆ 0.0         ┆ 1.0          ┆ 2.0          ┆ 1.0          │
└────────────┴────────────┴─────────────┴─────────────┴──────────────┴──────────────┴──────────────┘

Looks better

#### Counts of reviews and Count of text reviews

In [30]:
books_df['CountsOfReview'].describe()

shape: (6, 2)
┌────────────┬────────────┐
│ statistic  ┆ value      │
│ ---        ┆ ---        │
│ str        ┆ f64        │
╞════════════╪════════════╡
│ min        ┆ 0.0        │
│ max        ┆ 154447.0   │
│ null_count ┆ 0.0        │
│ mean       ┆ 11.580386  │
│ std        ┆ 295.280151 │
│ count      ┆ 1.850149e6 │
└────────────┴────────────┘

In [31]:
books_df['CountsOfReview'].value_counts()

shape: (2832, 2)
┌────────────────┬────────┐
│ CountsOfReview ┆ counts │
│ ---            ┆ ---    │
│ i64            ┆ u32    │
╞════════════════╪════════╡
│ 1552           ┆ 3      │
│ 40             ┆ 1185   │
│ 240            ┆ 48     │
│ 2504           ┆ 1      │
│ …              ┆ …      │
│ 1537           ┆ 1      │
│ 1769           ┆ 1      │
│ 1169           ┆ 1      │
│ 537            ┆ 6      │
└────────────────┴────────┘

In [32]:
books_df = books_df.rename({'Count of text reviews': 'CountOfTextReviews'})
books_df['CountOfTextReviews'].describe()

shape: (6, 2)
┌────────────┬────────────┐
│ statistic  ┆ value      │
│ ---        ┆ ---        │
│ str        ┆ f64        │
╞════════════╪════════════╡
│ min        ┆ 0.0        │
│ max        ┆ 43652.0    │
│ null_count ┆ 1.440462e6 │
│ mean       ┆ 7.172849   │
│ std        ┆ 124.840352 │
│ count      ┆ 1.850149e6 │
└────────────┴────────────┘